### PR TITLE
feat: add the json% macro from ProofWidgets4

### PR DESCRIPTION
**Pull request by PatrickMassot**
---

See Zulip discussion at https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/json.20elaborator

### CODE REVIEW

**Review by digama0 — CHANGES_REQUESTED** [review]
Comments/concerns:
* It should be in `Std` not `ProofWidgets`
* The instances at the start are misplaced, and should go in `Std.Lean` or in lean itself
* The syntax should not be scoped, and if this causes problems we need to fix those before merge (mathlib bump test needed)
* The `$(` token conflicts with lean antiquotations

**Comment by PatrickMassot** [comment]
Do you have a suggestion to replace `$(`?

**Comment by digama0** [comment]
ideally it would be `$(`, but some research is needed to determine how to do this without messing up other uses of antiquotations. Perhaps the implementation in Qq could be helpful...

**Comment by PatrickMassot** [comment]
I made the easy changes but the quotation thing may be above my pay grade.

**Comment by PatrickMassot** [comment]
@gebner do you have any insight about reusing the antiquotation syntax coming from Qq?

File: Std/Data/Json.lean [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+-/
+import Std.Lean.Json
+
+/-!
+# JSON-like syntax for Lean.
+
+Now you can write
+
+```lean
+open scoped ProofWidgets.Json
```
Comment (sgouezel) [inline Std/Data/Json.lean:14]:
```suggestion
open scoped Std.Json
```

**Comment by Vtec234** [comment]
Hello, I improved the antiquotation story [here](https://github.com/EdAyers/ProofWidgets4/commit/9f68f14df384f43b74aeb2908b97ae54a0ad9d95). Feel free to update the PR with those changes. I kept the PW4 variant `scoped` because even with the changes I am not confident that it won't break global syntax. Probably if you remove `scoped` here and both std4/mathlib4 still compile, then it's ok.

**Comment by gebner** [comment]
> do you have any insight about reusing the antiquotation syntax coming from Qq?

Qq uses the same quotation syntax as Lean core.  Wojciech has now changed the json macro in proofwidgets to do the same. ^^

**Comment by PatrickMassot** [comment]
I tried to incorporate changes from ProofWidgets.

File: Std/Data/Json.lean [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace Std.Json
+open Lean
+
+/-- Json data syntax. -/
+declare_syntax_cat jso (behavior := symbol)
+/-- Json data field syntax. -/
+declare_syntax_cat jso_field
+/-- Json identifier syntax. -/
+declare_syntax_cat jso_ident
```
Comment (Vtec234) [inline Std/Data/Json.lean:36]:
```suggestion
```

File: Std/Data/Json.lean [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+syntax "-"? num : jso
+/-- Json number negation syntax for scientific numbers. -/
+syntax "-"? scientific : jso
+/-- Json array syntax. -/
+syntax "[" jso,* "]" : jso
+/-- Json identifier syntax. -/
+syntax jsoIdent := ident <|> str
+/-- Json key/value syntax. -/
+syntax jsoField := jsoIdent ": " jso
+/-- Json dictionary syntax. -/
```
Comment (Vtec234) [inline Std/Data/Json.lean:55]:
```suggestion
/-- Json object syntax. -/
```

File: Std/Data/Json.lean [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  isBug : null,
+  lookACalc: $(23 + 54 * 2)
+}
+```
+-/
+
+namespace Std.Json
+open Lean
+
+/-- Json data syntax. -/
```
Comment (Vtec234) [inline Std/Data/Json.lean:31]:
```suggestion
/-- Json syntactic category -/
```

**Review by Vtec234 — APPROVED** [review]

**Comment by Vtec234** [comment]
@digama0 this is good to go I think.

File: Std/Lean/Json.lean [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    | none =>
+      match s with
+      | "inf" => "inf" -- [todo] emit a warning
+      | "-inf" => "-inf"
+      | "nan" => "nan"
+      | _ => panic! s!"unhandled float string {s}"
+    | some (m, e, de) =>
+      let j : JsonNumber := OfScientific.ofScientific m e de
+      let j := if x < 0.0 then -j else j
+      Json.num j
```
Comment (digama0) [inline Std/Lean/Json.lean:35]: This instance looks super hackish, translating stuff to strings and then back, and constructing invalid JSON literals like `nan`. What is going on here? (If you want to translate invalid floats I think the usual method is to use `null` instead.)

**Review by digama0 — APPROVED** [review]

**Pull request merged**